File: src/main/java/com/celeste/tasks/factory/TaskFactory.java
package com.celeste.tasks.factory;

import com.celeste.tasks.controller.TaskController;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class TaskFactory {

  private static final TaskFactory INSTANCE;

  static {
    INSTANCE = new TaskFactory();
  }

  public TaskController start() {
    final ExecutorService executor = ThreadingFactory.threadPool();

    return new TaskController(executor);
  }

  public TaskController start(final ExecutorService executor) {
    return new TaskController(executor);
  }

  public static TaskFactory getInstance() {
    return INSTANCE;
  }

}

File: src/main/kotlin/com/cefoler/tasks/factory/TaskFactory.kt
package com.cefoler.tasks.factory

import com.cefoler.tasks.controller.TaskController
import lombok.AccessLevel
import lombok.Getter
import lombok.NoArgsConstructor
import java.util.concurrent.ExecutorService

@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
class TaskFactory {

  fun create(): TaskController {
    val executor: ExecutorService = ThreadingFactory.threadPool()
    return TaskController(executor)
  }

  fun create(executor: ExecutorService): TaskController {
    return TaskController(executor)
  }

  companion object {
    val INSTANCE: TaskFactory = TaskFactory()
  }

}